File: 11. Stack/StackusingLinkedList.cpp
#include <bits/stdc++.h>
using namespace std;
 
typedef struct Node
{
    int data;
    struct Node* next;
} node;
 
node* top;

void push(int num)
{
    node* temp;
    temp=new Node();
    temp->data=num;
    temp->next=top;
    top=temp;
}

void pop()
{
    node* temp;
    if (top==NULL){
        cout<<"Stack Underflow\n";
        return;
    }
    else{
        temp=top;
        top=top->next;
        temp->next=NULL;
        free(temp);
    }
}

void display()
{
    node* temp;
    if (top==NULL){
        cout<<"No element in Stack\n";
        return;
    }
    else{
        temp=top;
        while (temp!=NULL){
            if(temp->next!=NULL){
                cout<<temp->data<<"-> ";
                temp=temp->next;
            }
            else{
                cout<<temp->data;
                temp=temp->next;
            }
        }
    }
}

int main()
{
    push(10);
    push(20);
    push(30);
    push(40);
    display();
    cout<<"\n";
    pop();
    pop();
    push(50);
    push(60);
    push(70);
    pop();
    push(80);
    display();
    cout<<"\n";
          
    return 0;
}

File: 11. Stack/stack-using-linked-list.py
# Here head is Top
class Node:
	def __init__(self, data):
		self.data = data
		self.next = None

class Stack:
	def __init__(self):
		self.head = None

	def isEmpty(self):
		if self.head == None:
			return True
		else:
			return False

	def push(self, data):
		if self.head == None:
			self.head = Node(data)
		else:
			newnode = Node(data)
			newnode.next = self.head
			self.head = newnode

	def popS(self):
		if self.isEmpty():
			print("Queue is empty.")
			return None
		else:
			temp = self.head
			self.head = self.head.next
			temp.next = None
			return temp.data

	def peek(self):
		if self.isEmpty():
			print("Queue is empty.")
			return None
		else:
			return self.head.data

	def display(self):
		if self.isEmpty():
			print("Queue is empty.")
		else:
			i = self.head
			print(i.data, "<--- Top")
			i = i.next
			while i != None:
				print(i.data)
				i = i.next

s = Stack()
s.push(1)
s.push(2)
s.push(3)
s.push(4)
s.push(5)
s.display()